### PR TITLE
Feature/validation

### PR DESCRIPTION
**Pull request by aroemen**
This is an extension to the previous pull request #17

### CODE REVIEW

**Review by mindsers — CHANGES_REQUESTED** [review]
Several comments of #17 apply this PR too. So, we'll need to check them there first.

**Comment by aroemen** [comment]
I'll pull these two PRs apart so they don't depend on each other.

**Comment by aroemen** [comment]
@mindsers I've cleaned up this PR to have it stand on its own as well.

**Comment by aroemen** [comment]
@mindsers I've rebased the changes you did from the merges on your master branch into this branch.

**Comment by mindsers** [comment]
@aroemen Can you rebase again the changes please?

**Comment by aroemen** [comment]
> @aroemen Can you rebase again the changes please?

@mindsers I've rebased this branch now.

**Comment by mindsers** [comment]
@aroemen Thank you so much for this huge bunch of work you did here. ❤️ 

I still have some thoughts/questions about it:
- I think this is a great feature to have but I don't want it to be a breaking change. So I would rather making it optional. My idea is: current default behavior is to read and make available the content of the CHANGELOG file (often needed on a CD pipelines where we don't want things to break) and for other use cases (like CI pipelines where we expect things to break) we can ask for validation and then the action acts like a CHANGELOG linter
- Can I ask you why the `validation_depth` is needed? I got what it does but do we want to validate only some entries?
- If we tend to act like a linter and since a lot of GitHub repo have multiple CHANGELOG, should we be able to read, parse and lint several files?

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - The `status` of the release based on the version number and the title line of the entry.
     Could be equal to `unreleased`, `prereleased`, `released` or `yanked`.
     Please refer to https://semver.org/#semantic-versioning-specification-semver for more informations about this.
+- Introduced changelog validation to help keep the release version in line with [Semantic Versioning](https://semver.org/)
+- New input param of `validation_depth` to allow for configuration of changelog validation.
```
Comment (mindsers) [inline CHANGELOG.md:18]: This CHANGELOG was badly merged, we should move those new lines to the unreleased section

**Review by mindsers — CHANGES_REQUESTED** [review]
Quick fix :)

**Comment by aroemen** [comment]
@mindsers I agree that having validation is more of a CI process for the most part and that this change shouldn't break the reading side of itself. Setting the validation depth property to 0 (which is the default value) prevents any validation from happening. So by default, this will not be a breaking change as the consumer would have to set this property in their workflow manually.

My thought behind the idea of the `validation_depth` property was to help people if they needed to transition to the KeepAChangelog format and for performance. Rather than forcing them to update their entire changelog, they could just start by updating a few entries and go from there. Also, I was worried that long changelogs could exponentially extend build times. Setting this property to a set value will prevent validation from occurring unnecessarily.

Reading and parsing multiple changelog files brings on another level of challenges. I don't see this as a necessary feature though at this time. Why would a build need to go way back in time to get release details of an old build? Also if a consumer wanted to parse/validate multiple changelogs, they could execute the action multiple times for each file.

File: CHANGELOG.md (hunk 1 above)
Comment (aroemen) [reply CHANGELOG.md]: @mindsers I've updated this

**Comment by mindsers** [comment]
> Reading and parsing multiple changelog files brings on another level of challenges. I don't see this as a necessary feature though at this time. Why would a build need to go way back in time to get release details of an old build? Also if a consumer wanted to parse/validate multiple changelogs, they could execute the action multiple times for each file.

Totally agree with you. Having a build step for each file is a better solution.

> I agree that having validation is more of a CI process for the most part and that this change shouldn't break the reading side of itself. Setting the validation depth property to 0 (which is the default value) prevents any validation from happening. So by default, this will not be a breaking change as the consumer would have to set this property in their workflow manually.
> 
> My thought behind the idea of the `validation_depth` property was to help people if they needed to transition to the KeepAChangelog format and for performance. Rather than forcing them to update their entire changelog, they could just start by updating a few entries and go from there. Also, I was worried that long changelogs could exponentially extend build times. Setting this property to a set value will prevent validation from occurring unnecessarily.

OK I understand why `validation_depth` is needed and that it already allows to NOT break existing build (thanks to its default value). But maybe we could think of an improved API. I mean, an API where the user ask explicitly for validation (for example a `validation` or `check` or `lint` boolean property). Then we could let `validation_depth` at a reasonable default value like 5 or 10, so user don't have to overthink what's a good value for it.

**Comment by mindsers** [comment]
Another question I have is what's about return value or logging message of a failed validation? Breaking a build saying "your changelog isn't valid" or breaking it saying " your changelog isn't valid because you put a breaking change in a minor release" is definitely not the same for the end user.

So, I think we should display better human readable error messages when the validation failed. What o you think?

**Comment by aroemen** [comment]
@mindsers I guess that is your call if you want an additional property or not. I didn't do it originally because I felt it was unnecessary to have additional properties. If you do add an additional property and the user wants to change how deep or shallow to validate, the consumers would now have to set multiple properties. Not the end of the world, but I'm always in the camp of less is more.

As for the error messaging, I think that can be added after the fact as an enhancement. While their really was no scope initially defined for this feature, this is additional development work on a PR that is already getting lengthy.

**Review by mindsers — APPROVED** [review]
I'm OK with that. Will work on it some more later. Thanks again for your work!! 😄

**Pull request merged**